it/src/main/java/com/google/cloud/teleport/it/pubsub/DefaultPubsubResourceManager.java

```diff
@@ -0,0 +1,218 @@
+  }
+
+  public static Builder builder(String testId, String projectId) {
+    checkArgument(!Strings.isNullOrEmpty(testId), "testId can not be null or empty");
+    checkArgument(!projectId.isEmpty(), "projectId can not be empty");
+    return new Builder(testId, projectId);
+  }
+
+  @Override
+  public TopicName createTopic(String topicName) {
+    checkArgument(!topicName.isEmpty(), "topicName can not be empty");
+
+    LOG.info("Creating topic '{}'", topicName);
+
+    Topic topic = topicAdminClient.createTopic(getTopicName(topicName));
+    TopicName reference = PubsubUtils.toTopicName(topic);
+    createdTopics.add(reference);
+
+    return reference;
+  }
+
+  @Override
+  public SubscriptionName createSubscription(String topicName, String subscriptionName) {
+    checkArgument(!topicName.isEmpty(), "topicName can not be empty");
+    checkArgument(!subscriptionName.isEmpty(), "subscriptionName can not be empty");
+
+    return createSubscription(getTopicName(topicName), subscriptionName);
+  }
+
+  @Override
+  public SubscriptionName createSubscription(TopicName topicName, String subscriptionName) {
+    checkArgument(!subscriptionName.isEmpty(), "subscriptionName can not be empty");
+
+    LOG.info("Creating subscription '{}' for topic '{}'", subscriptionName, topicName);
+
+    Subscription subscription =
+        subscriptionAdminClient.createSubscription(
+            getSubscriptionName(subscriptionName),
+            topicName,
+            PushConfig.getDefaultInstance(),
+            DEFAULT_ACK_DEADLINE_SECONDS);
+    SubscriptionName reference = PubsubUtils.toSubscriptionName(subscription);
+    createdSubscriptions.add(reference);
+
+    return reference;
+  }
+
+  @Override
+  public String publish(TopicName topic, Map<String, String> attributes, ByteString data)
+      throws PubsubResourceManagerException {
+    LOG.info("Publishing message with {} bytes to topic '{}'", data.size(), topic);
+
+    PubsubMessage pubsubMessage =
+        PubsubMessage.newBuilder().putAllAttributes(attributes).setData(data).build();
+
+    try {
+      String messageId = getPublisher(topic).publish(pubsubMessage).get();
+      LOG.info("Message published with id '{}'", messageId);
+      return messageId;
+    } catch (Exception e) {
+      throw new PubsubResourceManagerException("Error publishing message to Pubsub", e);
+    }
+  }
+
+  @VisibleForTesting
+  Publisher getPublisher(TopicName topic) throws IOException {
+    return Publisher.newBuilder(topic).setCredentialsProvider(credentialsProvider).build();
+  }
+
+  @Override
+  public synchronized void cleanupAll() {
+    LOG.info("Attempting to cleanup manager.");
+
+    try {
+      for (SubscriptionName subscription : createdSubscriptions) {
+        LOG.info("Deleting subscription '{}'", subscription);
+        subscriptionAdminClient.deleteSubscription(subscription);
+      }
+    } finally {
+      subscriptionAdminClient.close();
```

Comment:
Sorry, missed this before, but I think this puts the class in an unusable state. We should be checking that the clients haven't been closed and throwing an `IllegalStateException` if they are. (This method could just return, since we're in the expected final state.)

Reply:
Thanks. I've looked Spanner up and using `checkIsUsable` now.